src/main/java/cubicchunks/lighting/FirstLightProcessor.java

```diff
@@ -60,155 +56,198 @@
+
+					if (opacityIndex.isOccluded(localX, cubeMinY + localY, localZ)) {
+						break;
+					}
+
+					cube.setSkylight(localX, localY, localZ, 15);
+				}
 			}
-			cube.setLightFor(EnumSkyBlock.SKY, pos, 15);
-			return true;
-		});
+		}
 	}
 
-	public boolean diffuseSkylight(Cube generatedCube) {
-		ICubicWorld world = generatedCube.getWorld();
+	/**
+	 * Diffuses skylight in the given cube and all cubes affected by this update.
+	 *
+	 * @param cube the cube whose skylight is to be initialized
+	 */
+	public void diffuseSkylight(Cube cube) {
+		ICubicWorld world = cube.getWorld();
 
-		//cache min/max Y, generating them may be expensive
+		// Cache min/max Y, generating them may be expensive
 		int[][] minBlockYArr = new int[16][16];
 		int[][] maxBlockYArr = new int[16][16];
 
-		int minBlockX = cubeToMinBlock(generatedCube.getX());
-		int maxBlockX = cubeToMaxBlock(generatedCube.getX());
+		int minBlockX = Coords.cubeToMinBlock(cube.getX());
+		int maxBlockX = Coords.cubeToMaxBlock(cube.getX());
 
-		int minBlockZ = cubeToMinBlock(generatedCube.getZ());
-		int maxBlockZ = cubeToMaxBlock(generatedCube.getZ());
+		int minBlockZ = Coords.cubeToMinBlock(cube.getZ());
+		int maxBlockZ = Coords.cubeToMaxBlock(cube.getZ());
 
-		for (int blockX = minBlockX; blockX <= maxBlockX; blockX++) {
-			for (int blockZ = minBlockZ; blockZ <= maxBlockZ; blockZ++) {
-				Pair<Integer, Integer> minMax = getMinMaxLightUpdateY(cache, generatedCube, blockX, blockZ);
-				//if there is nothing to update - store contradicting data so we can detect it later
-				minBlockYArr[blockX - minBlockX][blockZ - minBlockZ] =
-						minMax == null ? Integer.MAX_VALUE : minMax.getLeft();
-				maxBlockYArr[blockX - minBlockX][blockZ - minBlockZ] =
-						minMax == null ? Integer.MIN_VALUE : minMax.getRight();
+		// Determine the block columns that require updating. If there is nothing to update, store contradicting data so
+		// we can skip the column later.
+		for (int localX = 0; localX <= Cube.SIZE - 1; ++localX) {
+			for (int localZ = 0; localZ <= Cube.SIZE - 1; ++localZ) {
+				Pair<Integer, Integer> minMax = getMinMaxLightUpdateY(cube, localX, localZ);
+				minBlockYArr[localX][localZ] = minMax == null ? Integer.MAX_VALUE : minMax.getLeft();
+				maxBlockYArr[localX][localZ] = minMax == null ? Integer.MIN_VALUE : minMax.getRight();
 			}
 		}
 
-		BlockPos.MutableBlockPos pos = mutablePos;
-		Int2ObjectMap<FastCubeBlockAccess> blockAccess =
-				new Int2ObjectOpenCustomHashMap<>(10, 0.75f, CUBE_Y_HASH);
+		Int2ObjectMap<FastCubeBlockAccess> blockAccessMap = new Int2ObjectOpenCustomHashMap<>(10, 0.75f, CUBE_Y_HASH);
 
-		Column column = generatedCube.getColumn();
+		Column column = cube.getColumn();
 		for (int blockX = minBlockX; blockX <= maxBlockX; blockX++) {
 			for (int blockZ = minBlockZ; blockZ <= maxBlockZ; blockZ++) {
-				pos.setPos(blockX, pos.getY(), blockZ);
+
+				this.mutablePos.setPos(blockX, this.mutablePos.getY(), blockZ);
```

Comment:
But you loose the knowledge that it is a field and not a local variable. Having it this way shows possible race-conditions immediately.